video/out/vo_drm.c

```diff
@@ -0,0 +1,509 @@
+{
+    struct priv *p = vo->priv;
+
+    vo->dwidth = p->device_w;
+    vo->dheight = p->device_h;
+    vo_get_src_dst_rects(vo, &p->src, &p->dst, &p->osd);
+
+    int32_t w = p->dst.x1 - p->dst.x0;
+    int32_t h = p->dst.y1 - p->dst.y0;
+
+    // p->osd contains the parameters assuming OSD rendering in window
+    // coordinates, but OSD can only be rendered in the intersection
+    // between window and video rectangle (i.e. not into panscan borders).
+    p->osd.w = w;
+    p->osd.h = h;
+    p->osd.mt = MPMIN(0, p->osd.mt);
+    p->osd.mb = MPMIN(0, p->osd.mb);
+    p->osd.mr = MPMIN(0, p->osd.mr);
+    p->osd.ml = MPMIN(0, p->osd.ml);
+
+    p->x = (p->device_w - w) >> 1;
+    p->y = (p->device_h - h) >> 1;
+
+    mp_sws_set_from_cmdline(p->sws, vo->opts->sws_opts);
+    p->sws->src = *params;
+    p->sws->dst = (struct mp_image_params) {
+        .imgfmt = IMGFMT_BGR0,
+        .w = w,
+        .h = h,
+        .d_w = w,
+        .d_h = h,
+    };
+
+    if (p->curframe) talloc_free(p->curframe);
+    p->curframe = mp_image_alloc(IMGFMT_BGR0, p->device_w, p->device_h);
+    mp_image_params_guess_csp(&p->sws->dst);
+    mp_image_set_params(p->curframe, &p->sws->dst);
+
+    if (mp_sws_reinit(p->sws) < 0)
+        return 1;
+
+    vo->want_redraw = true;
+    return 0;
+}
+
+static void draw_image(struct vo *vo, mp_image_t *mpi)
+{
+    struct priv *p = vo->priv;
+
+    struct modeset_buf *front_buf = &p->dev->bufs[p->dev->front_buf];
+
+    struct mp_rect src_rc = p->src;
+    src_rc.x0 = MP_ALIGN_DOWN(src_rc.x0, mpi->fmt.align_x);
+    src_rc.y0 = MP_ALIGN_DOWN(src_rc.y0, mpi->fmt.align_y);
+    mp_image_crop_rc(mpi, src_rc);
+
+    mp_sws_scale(p->sws, p->curframe, mpi);
+
+    osd_draw_on_image(vo->osd, p->osd, mpi ? mpi->pts : 0, 0, p->curframe);
+
+    int32_t shift = (p->device_w * p->y + p->x) * 4;
+    memcpy_pic(front_buf->map + shift,
+               p->curframe->planes[0],
+               (p->dst.x1 - p->dst.x0) * 4,
+               p->device_h,
+               p->device_w * 4,
+               p->curframe->stride[0]);
+
+    talloc_free(mpi);
+}
+
+static void flip_page(struct vo *vo)
+{
+    struct priv *p = vo->priv;
+
+    int ret = drmModeSetCrtc(p->fd, p->dev->crtc,
+                             p->dev->bufs[p->dev->front_buf].fb,
+                             0, 0, &p->dev->conn, 1, &p->dev->mode);
+    if (ret) {
+        MP_WARN(vo, "Cannot flip page for connector\n");
+    } else {
+        p->dev->front_buf ++;
+        p->dev->front_buf %= BUF_COUNT;
+    }
+}
+
+static int preinit(struct vo *vo)
+{
+    struct priv *p = vo->priv;
+    p->dev = NULL;
+    p->fd = 0;
+    p->old_crtc = NULL;
+    p->curframe = NULL;
+    p->sws = mp_sws_alloc(vo);
+
+    int ret;
+
+    ret = modeset_open(vo, &p->fd, p->device_path);
+    if (ret)
+        return ret;
+
+    ret = modeset_prepare_dev(vo, p->fd, p->connector_id, &p->dev);
+    if (ret)
+        return ret;
```

Comment:
preinit error codes are not necessarily the same as these functions use (mpv generally doesn't use the -errno convention). I'd just return -1 here.